modules/packages/Sort.chpl

```
@@ -802,6 +868,307 @@ proc selectionSort(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator)
+    const subbits = startbit + radixbits;
+    var nbigsubs = 0;
+    var bigsubs:[0..radix] (int,int);
+    const runningNow = here.runningTasks();
+
+    // Never recursively sort the first or last bins
+    // (these store the end)
+    // (TODO: we could if we wanted to break ties in a comparator)
+
+    for bin in 1..radix-1 {
+      // Does the bin contain more than one record?
+      const bin_start = offsets[bin];
+      const bin_end = if bin+1<=radix then offsets[bin+1]-1 else end_n;
+      const num = 1 + bin_end - bin_start;
+      if num <= 1 {
+        // do nothing
+      } else if num < settings.minForTask || runningNow >= settings.maxTasks {
+        // sort it in this thread
+        msbRadixSort(bin_start, bin_end, A, criterion, subbits, settings);
+      } else {
+        // Add it to the list of things to do in parallel
+        bigsubs[nbigsubs] = (bin_start, bin_end);
+        nbigsubs += 1;
+      }
+    }
+
+    forall (bin,(bin_start,bin_end)) in zip(0..#nbigsubs,bigsubs) {
+      msbRadixSort(bin_start, bin_end, A, criterion, subbits, settings);
+    }
+  } else {
+    // The serial version
+    for bin in 1..radix-1 {
+      // Does the bin contain more than one record?
+      const bin_start = offsets[bin];
+      const bin_end = if bin+1<=radix then offsets[bin+1]-1 else end_n;
+      const num = 1 + bin_end - bin_start;
+      if num > 1 {
+        // sort it in this thread
+        msbRadixSort(bin_start, bin_end, A, criterion, startbit + radixbits, settings);
+      }
+    }
+  }
+
+  if settings.CHECK_SORTS then checkSorted(start_n, end_n, A, criterion);
+}
+
+pragma "no doc"
+proc checkSorted(start_n:int, end_n:int, A:[], criterion, startbit = 0)
+{
+  for i in start_n+1..end_n {
+    var cmp = chpl_compare(A[i-1], A[i], criterion);
+    if cmp > 0 {
+      writeln("Error: not sorted properly at i=", i, " A[i-1]=", A[i-1], " A[i]=", A[i], " in start=", start_n, " end=", end_n);
+      writeln(A);
+      assert(false);
```

This seems like an odd way to fail. Maybe throw an error or halt?